### PR TITLE
solve_univariate_inequality: solving inequality with imaginary coefficients

### LINKED ISSUE #13105: solving inequality involving polynomial with imaginary coefficients fails
target new bahavior:

```
isolve = solve_univariate_inequality
eq = (x + I)*(x + 2*I) < 0
assert isolve(eq, x) == Eq(x, 0)  # this should not be False as in master
assert isolve(eq, x, relational=False) == FiniteSet(0)
x12 = (x - 1)*(x - 2)
assert isolve(eq.subs(x, x12), x) == Or(Eq(x, 1), Eq(x, 2))
```

Although the function is continuous for all reals, it doesn't produce reals except when the imaginary part vanishes. So there should be some additional processing to find the reals of interest for this corner case.

### PR DESCRIPTION
**Pull request by akash9712**
Added support for solving inequalities involving rational functions with
imaginary coefficients. solve_univariate_inequality() has been modified to
return those values of the generator for which the function assumes a real
value and satisfies the inequality. Tests have been added.

Examples:
```
>>> isolve = solve_univariate_inequality
>>> eq = (x + I)*(x + 2*I) < 0
>>> isolve(eq, x)
Eq(x, 0)
>>> x12 = (x - 1)*(x - 2)
>>> isolve(eq.subs(x, x12), x)
Eq(x, 1) | Eq(x, 2)
```

Fixes #13105

### CODE REVIEW

File: sympy/solvers/inequalities.py [hunk 1]
```diff
@@ -603,7 +623,11 @@ def valid(x):
             if valid(_pt(start, end)):
                 sol_sets.append(Interval.open(start, end))
 
-            rv = Union(*sol_sets).subs(gen, _gen)
+            if expanded_e.coeff(I) != S.Zero:
+                rv = (make_real)
+
```
Comment (ArifAhmed1995) [inline sympy/solvers/inequalities.py]: @akash9712 This extra blank line is not needed.

Comment (akash9712) [reply sympy/solvers/inequalities.py]: @ArifAhmed1995 I added that line to seperate the if clauses, as done before this pull request. Are you sure it should be removed?

Comment (ArifAhmed1995) [reply sympy/solvers/inequalities.py]: Yeah, I think blank line `628` should be removed.

**Comment by smichr** [comment]
This maybe should only be done if `expr.rel_op not in '!=='`; in case of Eq and Ne, imaginary solutions are permitted. What does this PR give for `isolve(Eq((x+I)*(x+2*I), -6), x)`? It should give `[-4*I, I]`.

**Comment by akash9712** [comment]
The PR would return `False` for isolve `(Eq((x+I)*(x+2*I), -6), x)`.
https://github.com/sympy/sympy/blob/master/sympy/solvers/inequalities.py#L449
This line suggests that the function should be called only for real domains. Should the function be changed to return complex values for the generator?

**Comment by smichr** [comment]
> Should the function be changed to return complex values for the generator?

No. Thanks for the reminder. solveset already solves the equation correctly. There is no need to call this function when dealing with an Eq instance.

**Comment by smichr** [comment]
Did you consider splitting the equation into real and imaginary parts rather than explicitly using "I"? What happens if you replace `I` with `Symbol("I", imaginary=True)`?

**Comment by akash9712** [comment]
I tried using the condition `im(expr) == 0` to check if the function contains an imaginary part, but the condition returns `False` in some cases even when the function would be real on the domain given by `continuous_domain`. For eg. 
```
>>> continuous_domain(log(x), x, S.Reals)
Interval.open(0, oo)
>>> im(log(x)) == 0
False
>>> im(log(x))
arg(x)
```

**Comment by akash9712** [comment]
@smichr I have tried working around this issue, but I needed to make certain changes to the `solveset` function for the solving of any function of type `arg`. Should I open a seperate pull request for that?

**Comment by akash9712** [comment]
@smichr Please review the changes.

File: sympy/solvers/inequalities.py [hunk 2]
```diff
@@ -571,6 +573,47 @@ def valid(x):
...
+                im_sol = FiniteSet()
+                try:
+                    a = solveset(im(expanded_e), gen, domain)
+                    if not isinstance(a, Interval):
+                        for z in a:
+                            if z not in singularities and valid(z) and z.is_real:
+                                im_sol += FiniteSet(z)
+                    else:
+                        start, end = a.inf, a.sup
+                        for z in _nsort(critical_points + FiniteSet(end)):
```
Comment (smichr) [inline sympy/solvers/inequalities.py]: Recommend that you `compute valid_start = valid(start)` once here after defining start and then use that variable below.

Comment (smichr) [reply sympy/solvers/inequalities.py]: Recommend that you compute `valid_z = valid(z)` immeadiately after starting the z loop.

File: sympy/solvers/inequalities.py [hunk 3]
```diff
@@ -571,6 +573,47 @@ def valid(x):
...
+                                if pt not in singularities and valid(pt) and pt.is_real:
+                                    if valid(start) and valid(z):
+                                        im_sol += Interval(start, z)
+                                    elif valid(start) and not valid(z):
+                                        im_sol += Interval.Ropen(start, z)
+                                    elif not valid(start) and valid(z):
+                                        im_sol += Interval.Lopen(start, z)
+                                    else:
+                                        im_sol += Interval.open(start, z)
+                            start = z
```
Comment (smichr) [inline sympy/solvers/inequalities.py]: `valid_start = valid_z`

**Comment by akash9712** [comment]
@smichr Made the changes as recommened.

**Comment by akash9712** [comment]
Thanks!

**Pull request merged**